docs/sources/lids/_DRAFT_QueryFairness.md

```diff
@@ -0,0 +1,319 @@
+FL -->> U1
+FL -->> U2
+FL -->> U3
+FL -->> U4
+
+package Querier {
+    [Worker] as W1
+    [Worker] as W2
+}
+
+T1 -->> Q1
+Q1 -->> W1
+
+T2 -->> Q1
+T2 -->> Q2
+Q2 -->> W2
+```
+
+![component diagram](../scheduler-proposal-1-component-diagram.png)
+
+Like the current implementation, the scheduler enqueues requests based on the `X-Scope-OrgID` header (or equivalent key in the request context), but also takes a second key (such as `X-Scope-UserID`) into account. This ensembles a fixed hierarchy with two levels where the tenant-to-user relation is a one-to-many relation.
+However, this has the disadvantage that the concept of users (that does not exist yet in Loki) leaks into the scheduler domain.
+
+**Pros:**
+* Relatively simple to to implement
+
+**Cons:**
+* Not extensible
+* Leaks domain knowledge
+
+### Proposal 2: Fully hierarchical scheduler
+
+This proposal is similar to _Proposal 1_, but with the difference that there are no fixed levels and levels can be nested arbitrarily.
+
+**Component diagram:**
+
+```plantuml
+skinparam componentStyle rectangle
+
+package Frontend {
+    () RoundTripper as QF
+}
+
+package Scheduler {
+    () FrontendLoop as FL
+
+    component RequestQueue {
+        [L3 Queue] as Z1
+        [L3 Queue] as Z2
+        [L3 Queue] as Z3
+        [L3 Queue] as Z4
+        [L2 Queue] as Y1
+        [L2 Queue] as Y2
+        [L2 Queue] as Y3
+        [L1 Queue] as X1
+        [L1 Queue] as X2
+        [L1 Queue] as X3
+        [L0 Queue] as R1
+        [L0 Queue] as R2
+    }
+
+    () QuerierLoop as Q1
+    () QuerierLoop as Q2
+
+    Z1 ..>> Y1
+    Z2 ..>> Y1
+    Z3 ..>> Y2
+    Z4 ..>> Y2
+    Y1 ..>> X1
+    Y2 ..>> X2
+    Y3 ..>> X2
+    X1 ..>> R1
+    X2 ..>> R2
+    X3 ..>> R2
+}
+
+QF -->> FL
+FL -->> Z1
+FL -->> Z2
+FL -->> Z3
+FL -->> Z4
+FL -->> Y3
+FL -->> X2
+FL -->> X3
+
+package Querier {
+    [Worker] as W1
+    [Worker] as W2
+}
+
+R1 -->> Q1
+Q1 -->> W1
+
+R2 -->> Q1
+R2 -->> Q2
+Q2 -->> W2
+```
+
+![component diagram](../scheduler-proposal-2-component-diagram.png)
+
+The implementation of the `RequestQueue`, which controls what querier workers are connected to which root queues (aka tenant queues), can be kept as is. However, the concept of tenants and users is dropped and replaced by by a concept of hierarchical actors, which can be represented as a slice of identifiers.
+
+**Example of identifiers:**
+
+```go
+actorA := []string{"tenant_a", "user_1"}
+actorB := []string{"tenant_b", "user_2"}
+actorC := []string{"tenant_b", "user_3", "service_foo"}
+actorD := []string{"tenant_b", "user_3", "service_bar"}
+```
+
+More generally:
+
+```go
+actorN := []string{"L0 Queue", "L1 Queue", "L2 Queue", ... "Ln Queue"}
+```
+
+The L0 queue (root queue) needs to be able to handle worker connections and therefore needs additional functionality compared to its leaf queues.
+
+The following code snippet is meant to show the simplified recursive structure of the queues.
+
+```go
+type RequestQueue struct {
```

Comment:
Could you add an example of `GetNextRequestForQuerier`. Also, should there be an interface `Queue`? This way you avoid `interface`.

Reply:
Yeah, I can add the interface. However, I don't want to be too specific in this document.